lib/ae_mdw_web/controllers/aex9_controller.ex

```diff
@@ -79,23 +79,26 @@ defmodule AeMdwWeb.Aex9Controller do
         )
 
   @spec balance_for_hash(Plug.Conn.t(), map()) :: Plug.Conn.t()
-  def balance_for_hash(conn, %{
+  def balance_for_hash(%Conn{assigns: %{state: state}} = conn, %{
         "blockhash" => block_hash_enc,
         "contract_id" => contract_id,
         "account_id" => account_id
-      }),
-      do:
-        handle_input(
-          conn,
-          fn ->
-            balance_for_hash_reply(
-              conn,
-              ensure_aex9_contract_pk!(contract_id),
-              Validate.id!(account_id, [:account_pubkey]),
-              ensure_block_hash_and_height!(block_hash_enc)
-            )
-          end
-        )
+      }) do
+    with {:ok, {type, height, hash}} <- ensure_block_hash(block_hash_enc),
+         {:ok, contract_pk} <- ensure_aex9_contract_at_block(state, contract_id, hash) do
+      handle_input(
+        conn,
+        fn ->
+          balance_for_hash_reply(
+            conn,
+            contract_pk,
+            Validate.id!(account_id, [:account_pubkey]),
```

Comment:
Perhaps we can start using `{:ok, account_pk} <- Validate.id(....) do` at the top to stop dealing with both error return values and exception errors?

It would be nice if we only dealt with just one of them, in particular, I think exceptions are very much error prone, so I would like to continue with the functional approach of returning an error tuple. Thoughts?

Reply:
I think the exceptions help with deeper propagation but for these validations close to the border I like the idea of refactoring incrementally to use tagged errors.